### PR TITLE
Feature: added pre commit hook to format and lint the code

### PR DESCRIPTION
**Pull request by HenriqueSFernandes**
Currently, the formatter (prettier) is running on **every** file of the project, regardless of whether there have been changes since the previous commit.

I will change this behavior in the future when everyone has already run the formatter to ensure all code is formatted, and then only staged files will be formatted.

### CODE REVIEW

**Comment by limwa** [comment]
It's okay to have the hook run on every file in the repo. Furthermore, with CI to ensure that every PR is properly formatted, no files outside of the ones touched in the PR will not be formatted. Having the hook run on every file instead of only on the files touched in the commit, at least in the case of uni, leads to some weird behavior when resolving merge conflicts.

File: pre-commit-hook.sh [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+# Define the pre-commit hook file path
+HOOKS_DIR=".git/hooks"
+PRE_COMMIT_FILE="$HOOKS_DIR/pre-commit"
+
+# Create the hooks directory if it doesn't exist
+mkdir -p $HOOKS_DIR
+
+# Write the pre-commit hook script
+cat <<'EOF' > $PRE_COMMIT_FILE
+#!/bin/bash
```
Comment (limwa) [inline pre-commit-hook.sh:12]:
```suggestion
#!/usr/bin/env bash
```
@tomaspalma nixos moment :rofl:

File: pre-commit-hook.sh [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+pnpm run format
+if [ $? -ne 0 ]; then
+  echo "Prettier formatting failed. Aborting commit."
+  exit 1
+fi
+
+# Stage any changes made by the formatter
+echo "Staging changes made by Prettier..."
+git add .
+
```
Comment (limwa) [inline pre-commit-hook.sh:29]:
```suggestion
```
In AdonisJS, prettier is integrated into ESLint. ESLint auto-fix also fixes Prettier issues.

File: pre-commit-hook.sh [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+pnpm run lint --fix
+if [ $? -ne 0 ]; then
+  echo "ESLint check failed. Aborting commit."
+  exit 1
+fi
+
+# Stage any changes made by the linter
+echo "Staging changes made by ESLint..."
+git add .
+
```
Comment (limwa) [inline pre-commit-hook.sh:41]: This hook doesn't take the following into account: there are 3 kinds of files in Git.

- Tracked and staged
- Tracked and unstaged
- Untracked

The same file can also be `tracked and staged` and `tracked and unstaged` at the same time (using `git add --patch`).
Therefore, this hook should only run on the `tracked and staged` patches.
To do that, you need to use commits and the stash to achieve the desired effect.

1. Before all linting, do `git stash push -S --quiet`. This ensures all staged changes (the one linting will be applied on) are stashed.
2. Then stash all other changed using `git stash push --include-untracked --quiet`.
3. Then pop the entry made in the first step using `git stash pop stash@{1}`.
4. Run all linting using `pnpm run lint --fix`.
5. Add all files using `git add .`
6. Then, unstash the changes that were stashed in 2 using `git stash pop`.
7. Then, proceed with the commit.

Please take into account that some `git stash pop` can cause conflicts and, thus, can cause this whole process to abort. What you should do in this case, for safety, is to revert to the initial state.

This can be done by creating an initial stash entry using `git stash push -u` and unstashing at the end using `git stash pop --index`.

**Review by limwa — CHANGES_REQUESTED** [review]

File: pre-commit-hook.sh (hunk 1 above)
Comment (tomaspalma) [reply pre-commit-hook.sh:12]: I learned how to write more portable bash scripts after changing to nixos and noticing no shell script for any repository that I tried was working on it :laughing:

**Comment by limwa** [comment]
Another option, which is also valid, is to do auto-formatting at the CI level and then only linting (without fixing) is performed at the commit level.

**Comment by HenriqueSFernandes** [comment]
# Major changes

Instead of a `pre-commit-hook.sh` file, there are 2 packages that are perfect for our use case:
- [Husky](https://typicode.github.io/husky/), which simplifies the creation of git hooks
- [lint-staged](https://github.com/lint-staged/lint-staged), which automatizes the process of linting only staged files, and rolling back in case of errors

In theory, after running `ppm install,` there is no need to run an additional command to register the hook.
Please test it to check if it works or if additional steps are required to register the hook.

**Comment by limwa** [comment]
This looks perfect! Will be checking this out ASAP!

File: website/.husky/pre-commit [hunk 4]
```diff
@@ -0,0 +1,2 @@
+cd website
+npx lint-staged
```
Comment (limwa) [inline website/.husky/pre-commit:2]:
```suggestion
pnpm lint-staged
```
Since our package manager for this project is pnpm, we can use it in the lint.
This also ensures that everyone is using the same version of pnpm (because of the packagemanager field in package.json)

File: website/package.json [hunk 5]
```diff
@@ -9,9 +9,10 @@
     "build": "node ace build",
     "dev": "node ace serve --hmr",
     "test": "node ace test",
-    "lint": "eslint .",
+    "lint": "eslint . --fix",
```
Comment (limwa) [inline website/package.json:12]: I don't like linting rules changing my code without me being aware of it, especially because, when I run `pnpm run lint`, this will run in the entire repository.

```suggestion
    "lint": "eslint .",
    "lint:fix": "eslint . --fix"
```

File: website/package.json [hunk 6]
```diff
@@ -44,13 +45,22 @@
     "@types/node": "^22.10.2",
     "@types/react": "^19.0.1",
     "@types/react-dom": "^19.0.2",
+    "@typescript-eslint/eslint-plugin": "^8.18.2",
+    "@typescript-eslint/parser": "^8.18.2",
     "@vitejs/plugin-react": "^4.3.4",
     "autoprefixer": "^10.4.20",
     "eslint": "^9.17.0",
+    "eslint-config-prettier": "^9.1.0",
+    "eslint-plugin-prettier": "^5.2.1",
```
Comment (limwa) [inline website/package.json:54]: These dependencies don't seem like they are needed. Aren't they already included i. @adonisjs/eslint_config?

File: website/package.json [hunk 7]
```diff
@@ -44,13 +45,22 @@
...
+    "eslint-plugin-react": "^7.37.3",
+    "eslint-plugin-react-hooks": "^5.1.0",
     "hot-hook": "^0.4.0",
+    "husky": "^9.1.7",
+    "lint-staged": "^15.2.11",
     "openapi-types": "^12.1.3",
     "pino-pretty": "^13.0.0",
     "prettier": "^3.4.2",
+    "prettier-edgejs": "^0.2.34",
```
Comment (limwa) [inline website/package.json:63]: Same as above

File: website/package.json [hunk 8]
```diff
@@ -141,5 +151,11 @@
     }
   },
   "prettier": "@adonisjs/prettier-config",
-  "packageManager": "pnpm@9.12.3+sha512.cce0f9de9c5a7c95bef944169cc5dfe8741abfb145078c0d508b868056848a87c81e626246cb60967cbd7fd29a6c062ef73ff840d96b3c86c40ac92cf4a813ee"
+  "packageManager": "pnpm@9.12.3+sha512.cce0f9de9c5a7c95bef944169cc5dfe8741abfb145078c0d508b868056848a87c81e626246cb60967cbd7fd29a6c062ef73ff840d96b3c86c40ac92cf4a813ee",
+  "eslintConfig": {
+    "extends": "@adonisjs/eslint-config/app"
+  },
```
Comment (limwa) [inline website/package.json:157]: Don't we have an eslint.config.js file for this?

File: website/package.json [hunk 9]
```diff
@@ -141,5 +151,11 @@
...
+  "lint-staged": {
+    "*": "pnpm run lint"
```
Comment (limwa) [inline website/package.json:159]:
```suggestion
    "*": "pnpm run lint:fix"
    "*": "pnpm run typecheck"
```

**Review by limwa — CHANGES_REQUESTED** [review]

File: website/.prettierignore [hunk 10]
```diff
@@ -1 +1,3 @@
 pnpm-lock.yaml
+build
+node_modules
```
Comment (limwa) [inline website/.prettierignore:3]: From the prettier documentation (https://prettier.io/docs/en/ignore.html)
> By default prettier ignores files in version control systems directories (".git", ".jj", ".sl", ".svn" and ".hg") and node_modules (unless the [--with-node-modules CLI option](https://prettier.io/docs/en/cli#--with-node-modules) is specified). Prettier will also follow rules specified in the ".gitignore" file if it exists in the same directory from which it is run.

As such, are these changes needed?
```suggestion
```

File: website/package.json [hunk 11]
```diff
@@ -48,9 +50,12 @@
     "autoprefixer": "^10.4.20",
     "eslint": "^9.17.0",
     "hot-hook": "^0.4.0",
+    "husky": "^9.1.7",
+    "lint-staged": "^15.2.11",
     "openapi-types": "^12.1.3",
     "pino-pretty": "^13.0.0",
     "prettier": "^3.4.2",
+    "prettier-edgejs": "^0.2.34",
```
Comment (limwa) [inline website/package.json:58]: This one is already included in `@adonisjs/prettier-config`

**Review by limwa — CHANGES_REQUESTED** [review]

File: website/package.json (hunk 11 above)
Comment (HenriqueSFernandes) [reply website/package.json:58]: Removing it from the dependencies causes this error:
```
Error: Cannot find package 'prettier-edgejs' imported from /home/ricky/Desktop/NIAEFEUP/enei/website/noop.js
    at __node_internal_ (file:///home/ricky/Desktop/NIAEFEUP/enei/website/node_modules/.pnpm/prettier@3.4.2/node_modules/prettier/index.mjs:15758:11)
    at new NodeError (file:///home/ricky/Desktop/NIAEFEUP/enei/website/node_modules/.pnpm/prettier@3.4.2/node_modules/prettier/index.mjs:15723:5)
    at packageResolve (file:///home/ricky/Desktop/NIAEFEUP/enei/website/node_modules/.pnpm/prettier@3.4.2/node_modules/prettier/index.mjs:16664:9)
    at moduleResolve (file:///home/ricky/Desktop/NIAEFEUP/enei/website/node_modules/.pnpm/prettier@3.4.2/node_modules/prettier/index.mjs:16704:18)
    at defaultResolve (file:///home/ricky/Desktop/NIAEFEUP/enei/website/node_modules/.pnpm/prettier@3.4.2/node_modules/prettier/index.mjs:16795:16)
    at resolve2 (file:///home/ricky/Desktop/NIAEFEUP/enei/website/node_modules/.pnpm/prettier@3.4.2/node_modules/prettier/index.mjs:16812:12)
    at importFromFile (file:///home/ricky/Desktop/NIAEFEUP/enei/website/node_modules/.pnpm/prettier@3.4.2/node_modules/prettier/index.mjs:16827:16)
    at importFromDirectory (file:///home/ricky/Desktop/NIAEFEUP/enei/website/node_modules/.pnpm/prettier@3.4.2/node_modules/prettier/index.mjs:21693:10)
    at importPlugin (file:///home/ricky/Desktop/NIAEFEUP/enei/website/node_modules/.pnpm/prettier@3.4.2/node_modules/prettier/index.mjs:21705:12)
    at async loadPluginWithoutCache (file:///home/ricky/Desktop/NIAEFEUP/enei/website/node_modules/.pnpm/prettier@3.4.2/node_modules/prettier/index.mjs:21709:18)
    at async Promise.all (index 0)
    at async Promise.all (index 1)
    at async Module.<anonymous> (file:///home/ricky/Desktop/NIAEFEUP/enei/website/node_modules/.pnpm/prettier@3.4.2/node_modules/prettier/index.mjs:21991:17)
    at async /home/ricky/Desktop/NIAEFEUP/enei/website/node_modules/.pnpm/eslint-plugin-prettier@5.2.1_eslint-config-prettier@9.1.0_eslint@9.17.0_jiti@1.21.6___eslint@_smc4lxph37ea4e7syguhj5l7sy/node_modules/eslint-plugin-prettier/worker.js:45:41
```

Comment (limwa) [reply website/package.json:58]: Did you run `pnpm install` after removing it from the dependencies?

Comment (HenriqueSFernandes) [reply website/package.json:58]: Yes

Comment (HenriqueSFernandes) [reply website/package.json:58]: It now works after I deleted the node_modules folder and reinstalled everything, weird

**Review by limwa — APPROVED** [review]

**Comment by limwa** [comment]
LGTM :rocket:
